File: flint/strategy/rsi.py
"""RSI (Relative Strength Index) mean-reversion strategy."""
from __future__ import annotations

from typing import List

import numpy as np

from ..models import Candle, Signal
from .base import Strategy


class RSIStrategy(Strategy):
    """Buy when RSI drops below oversold, sell when it rises above overbought."""

    def __init__(self, period: int = 14, oversold: float = 30, overbought: float = 70) -> None:
        self.period = period
        self.oversold = oversold
        self.overbought = overbought
        self._prev_rsi: float = 50.0

    @property
    def name(self) -> str:
        return f"RSI({self.period}, {self.oversold}/{self.overbought})"

    def reset(self) -> None:
        self._prev_rsi = 50.0

    def on_candle(self, candle: Candle, history: List[Candle]) -> Signal:
        if len(history) < self.period + 1:
            return Signal.HOLD

        closes = np.array([c.close for c in history[-(self.period + 1):]])
        deltas = np.diff(closes)
        gains = np.where(deltas > 0, deltas, 0)
        losses = np.where(deltas < 0, -deltas, 0)

        avg_gain = float(np.mean(gains))
        avg_loss = float(np.mean(losses))

        if avg_loss == 0:
            rsi = 100.0
        else:
            rs = avg_gain / avg_loss
            rsi = 100 - 100 / (1 + rs)

        signal = Signal.HOLD
        if self._prev_rsi >= self.oversold and rsi < self.oversold:
            signal = Signal.BUY
        elif self._prev_rsi <= self.overbought and rsi > self.overbought:
            signal = Signal.SELL

        self._prev_rsi = rsi
        return signal

File: ui/src/templates/rsi.ts
export const RSI_TEMPLATE = `from flint.strategy.base import Strategy
from flint.models import Candle, Signal, Side, TimeInForce
from typing import List
import numpy as np


class RSIMeanReversion(Strategy):
    """RSI Mean Reversion — buy oversold, sell overbought.

    Uses v2 execution: stop-loss, impact checks, proper sizing.
    Trend filter: only mean-revert in the direction of a 100-bar SMA.
    """
    def __init__(self, period=14, oversold=30, overbought=70, stop_pct=5.0, trend_period=100):
        self.period = period
        self.oversold = oversold
        self.overbought = overbought
        self.stop_pct = stop_pct / 100
        self.trend_period = trend_period

    @property
    def name(self) -> str:
        return f"RSI({self.period})"

    @classmethod
    def parameters(cls):
        return {
            "period": {"type": "int", "low": 5, "high": 30, "default": 14},
            "oversold": {"type": "float", "low": 15, "high": 40, "default": 30},
            "overbought": {"type": "float", "low": 60, "high": 85, "default": 70},
            "stop_pct": {"type": "float", "low": 2.0, "high": 10.0, "default": 5.0},
            "trend_period": {"type": "int", "low": 50, "high": 200, "default": 100},
        }

    def on_candle(self, candle: Candle, history: List[Candle], ctx=None) -> Signal:
        min_bars = max(self.period + 1, self.trend_period)
        if ctx is None or len(history) < min_bars:
            return Signal.HOLD
        closes = [c.close for c in history[-(self.period + 1):]]
        deltas = np.diff(closes)
        avg_gain = float(np.mean(np.where(deltas > 0, deltas, 0)))
        avg_loss = float(np.mean(np.where(deltas < 0, -deltas, 0)))
        rsi = 100 - (100 / (1 + avg_gain / avg_loss)) if avg_loss > 0 else 100

        # Trend filter: 100-bar SMA
        trend_closes = [c.close for c in history[-self.trend_period:]]
        sma = np.mean(trend_closes)
        uptrend = candle.close > sma

        if not ctx.positions:
            side = None
            if uptrend and rsi < self.oversold:
                side = Side.LONG
            elif not uptrend and rsi > self.overbought:
                side = Side.SHORT
            if side:
                size = (ctx.account.cash * 0.9) / candle.close
                impact = ctx.get_impact_price(candle.market, side, size)
                if impact and abs(impact - candle.close) / candle.close > 0.002:
                    return Signal.HOLD
                if size > 0:
                    ctx.market_order(candle.market, side, size)
                    stop_side = Side.SHORT if side == Side.LONG else Side.LONG
                    stop_price = candle.close * ((1 - self.stop_pct) if side == Side.LONG else (1 + self.stop_pct))
                    ctx.stop_order(candle.market, stop_side, size, stop_price)
        else:
            if 40 < rsi < 60:
                ctx.close_position(candle.market)
                ctx.cancel_all(candle.market)
        return Signal.HOLD

    def reset(self) -> None:
        pass
`
